seed/validate.go

```diff
@@ -24,62 +24,115 @@ import (
 	"fmt"
 	"path/filepath"
 	"regexp"
+	"sort"
 
-	"github.com/snapcore/snapd/seed/internal"
 	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/timings"
 )
 
+type ValidationError struct {
+	// SystemErrors maps system labels ("" for UC16/18) to their validation errors.
+	SystemErrors map[string][]error
+}
+
+func newValidationError(label string, err error) *ValidationError {
+	return &ValidationError{SystemErrors: map[string][]error{
+		label: {err},
+	}}
+}
+
+func (e *ValidationError) addErr(label string, errs ...error) {
+	if e.SystemErrors == nil {
+		e.SystemErrors = make(map[string][]error)
+	}
+	for _, err := range errs {
+		e.SystemErrors[label] = append(e.SystemErrors[label], err)
+	}
+}
+
+func (e ValidationError) hasErrors() bool {
+	return len(e.SystemErrors) != 0
+}
+
+func (e *ValidationError) Error() string {
+	systems := make([]string, 0, len(e.SystemErrors))
+	for s := range e.SystemErrors {
+		systems = append(systems, s)
+	}
+	sort.Strings(systems)
+	var buf bytes.Buffer
+	first := true
+	for _, s := range systems {
+		if first {
+			if s == "" {
+				fmt.Fprintf(&buf, "cannot validate seed:")
+			} else {
+				fmt.Fprintf(&buf, "cannot validate seed system %q:", s)
+			}
+		} else {
+			fmt.Fprintf(&buf, "\nand seed system %q:", s)
+		}
+		for _, err := range e.SystemErrors[s] {
+			fmt.Fprintf(&buf, "\n - %s", err)
+		}
+	}
+	return buf.String()
+}
+
 // ValidateFromYaml validates the given seed.yaml file and surrounding seed.
 func ValidateFromYaml(seedYamlFile string) error {
-	seed, err := internal.ReadSeedYaml(seedYamlFile)
+	// TODO:UC20: support validating also one or multiple UC20 seed systems
+	// introduce ListSystems ?
+	// What about full empty seed dir?
+	seedDir := filepath.Dir(seedYamlFile)
+
+	seed, err := Open(seedDir, "")
 	if err != nil {
-		return err
+		return newValidationError("", err)
+	}
+
+	if err := seed.LoadAssertions(nil, nil); err != nil {
+		return newValidationError("", err)
 	}
 
-	var errs []error
-	var haveCore, haveSnapd bool
+	tm := timings.New(nil)
```

Comment:
I'm really not sure how to do that, it's going to be called inside snapd and there might not even be a running snapd daemon around when it's running.

Reply:
what I was more referring to is that if we start using this command `snap debug validate-seed` automatically in our packaging scripts, it would be great if we had a way to keep tabs on that because it's entirely possible that the seeds that people start having on their systems grow large enough that this validation time becomes significant. 

I agree that I don't know how to do that here either considering the code is used as it is, but just something I wanted to bring up...